### PR TITLE
BatchDiagGGNExact Extension

### PR DESCRIPTION
**Pull request by sbharadwajj**
https://github.com/fKunstner/backpack-discuss/issues/67

Added `sum_batch` functionality for `DiagGGN` to get `BatchDiagGGN`

### CODE REVIEW

File: backpack/utils/conv.py [hunk 1]
```diff
@@ -47,19 +47,23 @@ def separate_channels_and_pixels(module, tensor):
...
     AX = einsum("nkl,vnml->vnkm", (input, grad_output_viewed))
-    weight_diagonal = (AX ** 2).sum([0, 1]).transpose(0, 1)
-    return weight_diagonal.view_as(module.weight)
+    sum_dims = [0, 1] if sum_batch else [0]
+    transpose_dims = (0, 1) if sum_batch else (1, 2)
+    weight_diagonal = (AX ** 2).sum(sum_dims).transpose(*transpose_dims)
+    if sum_batch:
+        return weight_diagonal.view_as(module.weight)
+    else:
+        return weight_diagonal.reshape(-1, *module.weight.numpy().shape)
```
Comment (f-dangel) [inline backpack/utils/conv.py:63]: remove the `numpy()`

Comment (f-dangel) [reply backpack/utils/conv.py:63]: Replace the `-1` by the more explicit `N = AX.shape[1]`

File: backpack/utils/conv_transpose.py [hunk 2]
```diff
@@ -35,12 +35,17 @@ def extract_weight_diagonal(module, input, grad_output, N):
...
-    weight_diagonal = (AX ** 2).sum([0, 1]).transpose(0, 1)
-    weight_diagonal = weight_diagonal.reshape(in_channels, out_channels, *k)
-    return weight_diagonal.transpose(0, 1)
+    sum_dims = [0, 1] if sum_batch else [0]
+    transpose_dims = (0, 1) if sum_batch else (1, 2)
+    weight_diagonal = (AX ** 2).sum(sum_dims).transpose(*transpose_dims)
+    if sum_batch:
+        weight_diagonal = weight_diagonal.reshape(in_channels, out_channels, *k)
+    else:
+        weight_diagonal = weight_diagonal.reshape(-1, in_channels, out_channels, *k)
```
Comment (f-dangel) [inline backpack/utils/conv_transpose.py:44]: Same as above

File: backpack/utils/linear.py [hunk 3]
```diff
@@ -1,9 +1,15 @@
...
+        return einsum("vno,ni->oi", (backproped ** 2, module.input0 ** 2))
+    else:
+        return einsum("vno,ni->noi", (backproped ** 2, module.input0 ** 2))
+
+
+def extract_bias_diagonal(module, backproped, sum_batch=True):
+    if sum_batch:
+        return einsum("vno->o", backproped ** 2)
+    else:
+        return einsum("vno->no", backproped ** 2)
```
Comment (f-dangel) [inline backpack/utils/linear.py:15]: Build the equation with `if` statements, then contract the tensors in a single call to `einsum`

File: requirements/test.txt [hunk 4]
```diff
@@ -5,3 +5,4 @@ pytest-optional-tests >= 0.1.1
 pytest-cov
 coveralls
 pytorch-memlab
+einops
```
Comment (f-dangel) [inline requirements/test.txt:8]: Try removing it, as it should already be installed through the main lib

File: test/extensions/implementation/base.py [hunk 5]
```diff
@@ -24,6 +24,10 @@ def diag_ggn(self):
         """Diagonal of Gauss Newton"""
         raise NotImplementedError
 
+    def diag_ggn_batch(self):
+        """Individual Diagonal of Gauss Newton with individual gradients"""
```
Comment (f-dangel) [inline test/extensions/implementation/base.py:28]: Remove `with individual gradients`

File: test/extensions/implementation/backpack.py [hunk 6]
```diff
@@ -69,6 +69,13 @@ def diag_ggn(self):
             diag_ggn = [p.diag_ggn_exact for p in self.problem.model.parameters()]
         return diag_ggn
 
+    def diag_ggn_batch(self):
+        with backpack(new_ext.BatchDiagGGNExact()):
+            _, _, loss = self.problem.forward_pass()
+            loss.backward()
+            diag_ggn_batch = [p.diag_ggn_batch for p in self.problem.model.parameters()]
```
Comment (f-dangel) [inline test/extensions/implementation/backpack.py:76]: `diag_ggn_exact_batch`

File: test/extensions/implementation/autograd.py [hunk 7]
```diff
@@ -66,27 +66,49 @@ def variance(self):
...
+                    diag_value = extract_ith_element_of_diag_ggn(
+                        parameter_index, p, loss, output
+                    )
+                    diag_ggn_p[parameter_index] = diag_value
+
+                diag_ggns.append(diag_ggn_p.view(p.size()))
+            return diag_ggns
+
+        _, batch_output, batch_loss = self.problem.forward_pass()
+        loss_list = torch.zeros((batch_size))
```
Comment (f-dangel) [inline test/extensions/implementation/autograd.py:95]: Create on same device as loss

File: test/extensions/implementation/backpack.py [hunk 8]
```diff
@@ -69,6 +69,13 @@ def diag_ggn(self):
             diag_ggn = [p.diag_ggn_exact for p in self.problem.model.parameters()]
         return diag_ggn
 
+    def diag_ggn_batch(self):
```
Comment (f-dangel) [inline test/extensions/implementation/backpack.py:72]: `diag_ggn_exact_batch`

File: test/extensions/secondorder/diag_ggn/test_diag_ggn.py [hunk 9]
```diff
@@ -21,7 +21,23 @@ def test_diag_ggn(problem):
     problem.set_up()
 
     backpack_res = BackpackExtensions(problem).diag_ggn()
-    autograd_res = AutogradExtensions(problem).diag_ggn()
+    autograd_res = AutogradExtensions(problem).diag_ggn(sum_batch=True)
```
Comment (f-dangel) [inline test/extensions/secondorder/diag_ggn/test_diag_ggn.py:24]: There should be two functions: `diag_ggn`, `diag_ggn_batch`

File: test/extensions/secondorder/diag_ggn/test_diag_ggn.py [hunk 10]
```diff
@@ -21,7 +21,23 @@ def test_diag_ggn(problem):
...
+
+    check_sizes_and_values(autograd_res, backpack_res)
+    problem.tear_down()
+
+
+@pytest.mark.parametrize("problem", PROBLEMS, ids=IDS)
+def test_diag_ggn_batch(problem):
+    """Test the diagonal of Gauss-Newton
```
Comment (f-dangel) [inline test/extensions/secondorder/diag_ggn/test_diag_ggn.py:32]: Update docstring

File: backpack/extensions/secondorder/diag_ggn/__init__.py [hunk 11]
```diff
@@ -114,6 +114,57 @@ def __init__(self):
...
+
+    For a faster but less precise alternative,
+    see :py:meth:`backpack.extensions.DiagGGNMC`.
+
+    """
+
+    def __init__(
+        self,
+        loss_hessian_strategy=LossHessianStrategy.EXACT,
+        savefield="diag_ggn_batch",
```
Comment (f-dangel) [inline backpack/extensions/secondorder/diag_ggn/__init__.py:133]: `diag_ggn_exact_batch`

File: backpack/extensions/secondorder/diag_ggn/__init__.py [hunk 12]
```diff
@@ -114,6 +114,57 @@ def __init__(self):
         )
 
 
+class BatchDiagGGNExact(BackpropExtension):
```
Comment (f-dangel) [inline backpack/extensions/secondorder/diag_ggn/__init__.py:117]: There should be an abstract class `BatchDiagGGN`, which is the parent of the `Exact` and `MC` extensions

File: backpack/extensions/secondorder/diag_ggn/__init__.py [hunk 13]
```diff
@@ -114,6 +114,57 @@ def __init__(self):
...
+    """
+    Individual Diagonal of the Generalized Gauss-Newton/Fisher.
```
Comment (f-dangel) [inline backpack/extensions/secondorder/diag_ggn/__init__.py:119]: Update docstring

File: backpack/extensions/secondorder/diag_ggn/__init__.py [hunk 14]
```diff
@@ -114,6 +114,57 @@ def __init__(self):
...
+    Uses the exact Hessian of the loss w.r.t. the model output.
+
+    Stores the output in :code:`diag_ggn_batch`,
```
Comment (f-dangel) [inline backpack/extensions/secondorder/diag_ggn/__init__.py:122]: Update docstring

File: backpack/extensions/__init__.py [hunk 15]
```diff
@@ -28,7 +30,9 @@
     "KFRA",
     "HBP",
     "DiagGGNExact",
+    "BatchDiagGGNExact",
     "DiagGGNMC",
     "DiagGGN",
+    "BatchDiagGGN",
```
Comment (f-dangel) [inline backpack/extensions/__init__.py:36]: Since `BatchDiagGGN` is the abstract parent class of `BatchDiagGGNExact` (and in the future `BatchDiagGGNMC`) it may not be required here.

File: backpack/extensions/secondorder/__init__.py [hunk 16]
```diff
@@ -18,14 +18,16 @@
...
+from .diag_ggn import DiagGGN, DiagGGNExact, BatchDiagGGN, BatchDiagGGNExact, DiagGGNMC
 from .diag_hessian import DiagHessian
 from .hbp import HBP, KFAC, KFLR, KFRA
 
 __all__ = [
     "DiagGGNExact",
+    "BatchDiagGGNExact",
     "DiagGGNMC",
     "DiagGGN",
+    "BatchDiagGGN",
```
Comment (f-dangel) [inline backpack/extensions/secondorder/__init__.py:30]: Same as in the remark above of the parent-class role of `BatchDiagGGN`

File: backpack/extensions/secondorder/diag_ggn/__init__.py [hunk 17]
```diff
@@ -136,3 +136,83 @@ def __init__(self, mc_samples=1):
 
     def get_num_mc_samples(self):
         return self._mc_samples
+
+
+class BatchDiagGGN(BackpropExtension):
+    """
+    Individual Diagonal of the Generalized Gauss-Newton/Fisher.
```
Comment (f-dangel) [inline backpack/extensions/secondorder/diag_ggn/__init__.py:143]: make "Diagonal" lower-case

File: backpack/extensions/secondorder/diag_ggn/__init__.py [hunk 18]
```diff
@@ -136,3 +136,83 @@ def __init__(self, mc_samples=1):
...
+    Uses the exact Hessian of the loss w.r.t. the model output.
+
+    Stores the output in :code:`diag_ggn_exact_batch`,
+    has the same dimensions as the gradient.
```
Comment (f-dangel) [inline backpack/extensions/secondorder/diag_ggn/__init__.py:147]: This statement is not correct. In fact, the output has the same dimension as the individual gradient. Can you look up the formulation of the output shape in the documentation of `BatchGrad` and insert it here?

File: backpack/extensions/secondorder/diag_ggn/__init__.py [hunk 19]
```diff
@@ -136,3 +136,83 @@ def __init__(self, mc_samples=1):
...
+
+    For a faster but less precise alternative,
+    see :py:meth:`backpack.extensions.DiagGGNMC`.
```
Comment (f-dangel) [inline backpack/extensions/secondorder/diag_ggn/__init__.py:150]: `DiagGGNMC` is not a cheap approximation of `BatchDiagGGNExact`

File: backpack/extensions/secondorder/diag_ggn/__init__.py [hunk 20]
```diff
@@ -136,3 +136,83 @@ def __init__(self, mc_samples=1):
...
+    Diagonal of the Generalized Gauss-Newton/Fisher.
+    Uses the exact Hessian of the loss w.r.t. the model output.
+
+    Stores the output in :code:`diag_ggn_exact`,
+    has the same dimensions as the gradient.
+
+    For a faster but less precise alternative,
+    see :py:meth:`backpack.extensions.DiagGGNMC`.
+
+    """
```
Comment (f-dangel) [inline backpack/extensions/secondorder/diag_ggn/__init__.py:212]: I added remarks above how to refactor the docstring of `BatchDiagGGN`. After that, move the docstring from `BatchDiagGGN` to this place such that `BatchDiagGGN`'s docstring is empty.

File: backpack/utils/conv_transpose.py [hunk 21]
```diff
@@ -33,14 +33,21 @@ def extract_weight_diagonal(module, input, grad_output, N):
 
     input_reshaped = input.reshape(M, -1, spatial_out_numel)
 
+    V_axis, N_axis = 0, 1
     grad_output_viewed = separate_channels_and_pixels(module, grad_output)
     AX = einsum("nkl,vnml->vnkm", (input_reshaped, grad_output_viewed))
-    weight_diagonal = (AX ** 2).sum([0, 1]).transpose(0, 1)
-    weight_diagonal = weight_diagonal.reshape(in_channels, out_channels, *k)
-    return weight_diagonal.transpose(0, 1)
+    N = AX.shape[1]
```
Comment (f-dangel) [inline backpack/utils/conv_transpose.py:39]: Replace `1` by `N_axis`

File: backpack/utils/conv.py [hunk 22]
```diff
@@ -47,19 +47,25 @@ def separate_channels_and_pixels(module, tensor):
...
     input must be the unfolded input to the convolution (see unfold_func)
     and grad_output the backpropagated gradient
     """
-
+    V_axis, N_axis = 0, 1
     grad_output_viewed = separate_channels_and_pixels(module, grad_output)
     AX = einsum("nkl,vnml->vnkm", (input, grad_output_viewed))
-    weight_diagonal = (AX ** 2).sum([0, 1]).transpose(0, 1)
-    return weight_diagonal.view_as(module.weight)
+    N = AX.shape[1]
```
Comment (f-dangel) [inline backpack/utils/conv.py:58]: Replace `1` by `N_axis`

File: test/extensions/implementation/autograd.py [hunk 23]
```diff
@@ -66,28 +66,49 @@ def variance(self):
...
-        def extract_ith_element_of_diag_ggn(i, p):
+        def extract_ith_element_of_diag_ggn(i, p, loss, output):
             v = torch.zeros(p.numel()).to(self.problem.device)
             v[i] = 1.0
             vs = vector_to_parameter_list(v, [p])
             GGN_vs = ggn_vector_product_from_plist(loss, output, [p], vs)
             GGN_v = torch.cat([g.detach().view(-1) for g in GGN_vs])
             return GGN_v[i]
 
-        diag_ggns = []
```
Comment (f-dangel) [inline test/extensions/implementation/autograd.py:80]: This is an unnecessary diff

File: test/extensions/implementation/autograd.py [hunk 24]
```diff
@@ -66,28 +66,49 @@ def variance(self):
...
 
+    def diag_ggn(self):
+        _, output, loss = self.problem.forward_pass()
+        return self._get_diag_ggn(loss, output)
+
+    def diag_ggn_batch(self):
+        batch_size = self.problem.input.shape[0]
+        _, _, batch_loss = self.problem.forward_pass()
+        batch_loss_device = batch_loss.device
+        loss_list = torch.zeros((batch_size), device=batch_loss_device)
```
Comment (f-dangel) [inline test/extensions/implementation/autograd.py:100]: I think you can replace `(batch_size)` with `batch_size` and achieve the same

Comment (f-dangel) [reply test/extensions/implementation/autograd.py:100]: I also saw that we have access to the `device` via `self.problem.device`. You can replace `batch_loss_device` and save one line.

File: test/extensions/implementation/base.py [hunk 25]
```diff
@@ -24,6 +24,10 @@ def diag_ggn(self):
         """Diagonal of Gauss Newton"""
         raise NotImplementedError
 
+    def diag_ggn_batch(self):
+        """Individual Diagonal of Gauss Newton"""
```
Comment (f-dangel) [inline test/extensions/implementation/base.py:28]: Diagonal -> diagonal
Gauss Newton -> Generalized Gauss-Newton/Fisher

File: test/extensions/secondorder/diag_ggn/test_diag_ggn.py [hunk 26]
```diff
@@ -27,6 +27,22 @@ def test_diag_ggn(problem):
     problem.tear_down()
 
 
+@pytest.mark.parametrize("problem", PROBLEMS, ids=IDS)
+def test_diag_ggn_batch(problem):
+    """Test the individual diagonal of Gauss-Newton
```
Comment (f-dangel) [inline test/extensions/secondorder/diag_ggn/test_diag_ggn.py:32]: same as above

File: backpack/extensions/secondorder/__init__.py (hunk 16 above)
Comment (f-dangel) [inline backpack/extensions/secondorder/__init__.py:30]: Similar to `extensions.__init__`, this import can be removed.

File: backpack/extensions/secondorder/diag_ggn/__init__.py [hunk 27]
```diff
@@ -136,3 +136,67 @@ def __init__(self, mc_samples=1):
...
+    ]
+
+    def __init__(self, loss_hessian_strategy=LossHessianStrategy.EXACT, savefield=None):
+        if savefield is None:
+            savefield = "diag_ggn_batch"
+        if loss_hessian_strategy not in self.VALID_LOSS_HESSIAN_STRATEGIES:
+            raise ValueError(
+                "Unknown hessian strategy: {}".format(loss_hessian_strategy)
+                + "Valid strategies: [{}]".format(self.VALID_LOSS_HESSIAN_STRATEGIES)
+            )
```
Comment (f-dangel) [inline backpack/extensions/secondorder/diag_ggn/__init__.py:154]: Since this is the parent class, `savefield` and `loss_hessian_strategy` should be set through the child constructor. Can you make both arguments non-keyword? I know this is different in `DiagGGN`, but not having default values for both fields better reflects the role of `BatchDiagGGN` as abstract base class

File: backpack/utils/conv_transpose.py [hunk 28]
```diff
@@ -33,14 +33,21 @@ def extract_weight_diagonal(module, input, grad_output, N):
...
     AX = einsum("nkl,vnml->vnkm", (input_reshaped, grad_output_viewed))
-    weight_diagonal = (AX ** 2).sum([0, 1]).transpose(0, 1)
-    weight_diagonal = weight_diagonal.reshape(in_channels, out_channels, *k)
-    return weight_diagonal.transpose(0, 1)
+    N = AX.shape[N_axis]
+    sum_dims = [V_axis, N_axis] if sum_batch else [V_axis]
+    transpose_dims = (V_axis, N_axis) if sum_batch else (V_axis + 1, N_axis + 1)
+    weight_diagonal = (AX ** 2).sum(sum_dims).transpose(*transpose_dims)
+    if sum_batch:
+        weight_diagonal = weight_diagonal.reshape(in_channels, out_channels, *k)
```
Comment (f-dangel) [inline backpack/utils/conv_transpose.py:44]: Can you change that to `view_as(module.weight)`, similar to `extract_weight_diagonal` in `utils/conv.py`?

File: backpack/utils/conv_transpose.py [hunk 29]
```diff
@@ -33,14 +33,21 @@ def extract_weight_diagonal(module, input, grad_output, N):
...
+    else:
+        weight_diagonal = weight_diagonal.reshape(N, in_channels, out_channels, *k)
```
Comment (f-dangel) [inline backpack/utils/conv_transpose.py:46]: `in_channels, out_channels, *k` can be replaced with `module.weight.shape`

File: test/extensions/implementation/autograd.py [hunk 30]
```diff
@@ -82,12 +80,33 @@ def extract_ith_element_of_diag_ggn(i, p):
...
+
+        batch_diag_ggn = []
+        for b in range(batch_size):
+            _, output, loss = self.problem.forward_pass(sample_idx=b)
+            diag_ggn = self._get_diag_ggn(loss, output)
+            batch_diag_ggn.append(diag_ggn)
+            loss_list[b] = loss
+        factor = self.problem.get_reduction_factor(batch_loss, loss_list)
+        params_batch_diag_ggn = list(zip(*batch_diag_ggn))
+        return [torch.stack(param) * factor for param in params_batch_diag_ggn]
```
Comment (f-dangel) [inline test/extensions/implementation/autograd.py:108]: Can you add a comment on the data layout?

Something like:

`batch_diag_ggn` has entries `[sample_idx][param_idx]`
`params_batch_diag_ggn` has entries `[param_idx][sample_idx]`

File: backpack/extensions/secondorder/diag_ggn/__init__.py [hunk 31]
```diff
@@ -136,3 +136,67 @@ def __init__(self, mc_samples=1):
...
+        )
+
+
+class BatchDiagGGNExact(BatchDiagGGN):
+    """
+    Individual diagonal of the Generalized Gauss-Newton/Fisher.
+    Uses the exact Hessian of the loss w.r.t. the model output.
+
+    Stores the output in ``diag_ggn_exact_batch`` as a ``[N x ...]`` tensor,
+    where ``N`` batch size and ``...`` is the shape of the gradient.
```
Comment (f-dangel) [inline backpack/extensions/secondorder/diag_ggn/__init__.py:195]: Typo: ``N`` is the batch size

File: backpack/utils/conv_transpose.py (hunk 29 above)
Comment (sbharadwajj) [reply backpack/utils/conv_transpose.py:46]: Hi, we cant do this because we are shifting the channels for `conv_transpose`

File: backpack/utils/conv_transpose.py (hunk 28 above)
Comment (sbharadwajj) [reply backpack/utils/conv_transpose.py:44]: This also does work with `view_as(module.weight)`, I have the error message

File: test/extensions/implementation/autograd.py (hunk 30 above)
Comment (sbharadwajj) [reply test/extensions/implementation/autograd.py:112]: I quite don't understand this comment, I have added exactly the same. But would it be better with shapes?

File: backpack/utils/conv_transpose.py (hunk 28 above)
Comment (f-dangel) [reply backpack/utils/conv_transpose.py:44]: The intermediate result is transposed, hence the target shape is not `model.shape`

File: backpack/extensions/secondorder/diag_ggn/__init__.py [hunk 32]
```diff
@@ -132,3 +132,59 @@ def __init__(self, mc_samples=1):
...
+                MSELoss: losses.DiagGGNMSELoss(),
+                CrossEntropyLoss: losses.DiagGGNCrossEntropyLoss(),
+                Linear: linear.BatchDiagGGNLinear(),
+                MaxPool1d: pooling.DiagGGNMaxPool1d(),
+                MaxPool2d: pooling.DiagGGNMaxPool2d(),
+                AvgPool1d: pooling.DiagGGNAvgPool1d(),
+                MaxPool3d: pooling.DiagGGNMaxPool3d(),
+                AvgPool2d: pooling.DiagGGNAvgPool2d(),
+                AvgPool3d: pooling.DiagGGNAvgPool3d(),
+                ZeroPad2d: padding.DiagGGNZeroPad2d(),
```
Comment (f-dangel) [inline backpack/extensions/secondorder/diag_ggn/__init__.py:157]: Can we recycle those classes, or will they fetch and backpropagate the 'wrong' quantity?

File: backpack/extensions/secondorder/diag_ggn/__init__.py [hunk 33]
```diff
@@ -132,3 +132,59 @@ def __init__(self, mc_samples=1):
...
+                ConvTranspose3d: convtranspose3d.BatchDiagGGNConvTranspose3d(),
+                Dropout: dropout.DiagGGNDropout(),
+                Flatten: flatten.DiagGGNFlatten(),
+                ReLU: activations.DiagGGNReLU(),
+                Sigmoid: activations.DiagGGNSigmoid(),
+                Tanh: activations.DiagGGNTanh(),
+                LeakyReLU: activations.DiagGGNLeakyReLU(),
+                LogSigmoid: activations.DiagGGNLogSigmoid(),
+                ELU: activations.DiagGGNELU(),
+                SELU: activations.DiagGGNSELU(),
```
Comment (f-dangel) [inline backpack/extensions/secondorder/diag_ggn/__init__.py:172]: Can we recycle those classes, or will they fetch and backpropagate the 'wrong' quantity?

File: backpack/extensions/secondorder/diag_ggn/__init__.py [hunk 34]
```diff
@@ -132,3 +132,59 @@ def __init__(self, mc_samples=1):
...
+
+
+class BatchDiagGGN(BackpropExtension):
+    VALID_LOSS_HESSIAN_STRATEGIES = [
+        LossHessianStrategy.EXACT,
+        LossHessianStrategy.SAMPLING,
+    ]
+
+    def __init__(self, loss_hessian_strategy, savefield):
+        self.loss_hessian_strategy = loss_hessian_strategy
```
Comment (f-dangel) [inline backpack/extensions/secondorder/diag_ggn/__init__.py:144]: Could you introduce a check for `loss_hessian_strategy` similar to [here](https://github.com/f-dangel/backpack/blob/development/backpack/extensions/secondorder/diag_ggn/__init__.py#L58-L64)?

Comment (sbharadwajj) [reply backpack/extensions/secondorder/diag_ggn/__init__.py:144]: hi, in a previous comment you asked me to remove this because it makes more sense not to have default values.

Comment (f-dangel) [reply backpack/extensions/secondorder/diag_ggn/__init__.py:144]: That's true. But I am referring to reintroducing the check, rather than the default value

Comment (f-dangel) [reply backpack/extensions/secondorder/diag_ggn/__init__.py:152]: Added
